Add GitHub reusable workflows for PR preview via surge

tlylt opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"

  If the pull request completely addresses the issue, use one of the issue closing keywords. https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Fixes #1466

Provide three convenient reusable workflows that users who are doing a forking workflow can use to set up PR preview via surge.sh easily. Some context and rationale is [here](https://github.com/MarkBind/markbind/issues/1466#issuecomment-1064043315)
1. fork-build: Build site from Fork
2. fork-preview: Deploy the Preview site from Fork to surge
3. unpublish-preview: (Optional) Teardown surge site after PR is closed

**Anything you'd like to highlight/discuss:**
- The new GitHub Reusable Workflows should probably reside in MarkBind, either as a new stand-alone 'markbind-reusable-workflows' repo or within the markbind repo itself. Due to the nature of the reusable workflows, they will have to reside in `.github/workflows`.
  - In view of the overhead of maintenance and release of a new repo, I would opt to put them together in the `MarkBind/markbind` repo, since unlike typical GitHub Actions, these workflows won't be released in the marketplace. Hence, can be managed together with the main markbind.
- There is one issue with site.json reported in https://github.com/MarkBind/markbind/issues/1825, also affects this workflow
- Unlike Actions, there is no need to publish reusable workflows. Testing can be done via e.g 'tlylt/markbind-reusable-workflows/.github/workflows/fork-build.yml@main'. See [readme](https://github.com/tlylt/markbind-reusable-workflows#usage) for more details.
- For the inputs to the workflow, there is one major difference between this and the markbind-action: `baseUrl` option is removed since there's no utility to set it when using Surge (unlike gh-pages, surge allows a domain from root)

**Testing instructions:**
Mainly three things to test
- PRs from the same repo should trigger the workflows
- PRs from forked repos should also trigger the workflows
- PRs that are closed can trigger the teardown of Surge site

#### Sample test runs
**PR from the same repo but different branches https://github.com/tlylt/test-md/pull/8**
- Create a PR from a branch of the same repo will trigger [build](https://github.com/tlylt/test-md/actions/runs/2145919994) and [preview](https://github.com/tlylt/test-md/actions/runs/2145922528)
- Update a PR from a branch of the same repo will re-trigger [build](https://github.com/tlylt/test-md/actions/runs/2145928065) and [preview](https://github.com/tlylt/test-md/actions/runs/2145931532)
- Close PR will trigger [unpublish](https://github.com/tlylt/test-md/actions/runs/2145938071)

**PR from a fork repo**
- Create a PR from a branch of the fork repo will trigger [build](https://github.com/tlylt/test-md/actions/runs/2146005541) and [preview](https://github.com/tlylt/test-md/actions/runs/2146006932)
- Update a PR from a branch of the fork repo will re-trigger [build](https://github.com/tlylt/test-md/actions/runs/2146016793) and [preview](https://github.com/tlylt/test-md/actions/runs/2146019353)
- Close PR will trigger [unpublish](https://github.com/tlylt/test-md/actions/runs/2146034492)

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Add utility GitHub reusable workflows for PR preview

<!--
  See this link for more info on how to write a good commit message:
  https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

  As best as possible, write a succinct commit title in 50 characters
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [x] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes <!-- Its tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

jonahtanjz commented:
> However, if the site is published via GitHub Pages, then users will need to explicitly specify baseUrl='' in this workflow to override the relative url, as PR preview is via Surge.sh, without the need for a relative URL(Doesn't seem to work with a relative URL either).

The fork-build workflow can be reused for other PR previewing platforms too right other than surge? If that is the case then maybe we can try to explore the other popular platforms to see if the default relative url is `""` or `"./repo_name"`. If most of them are `""`, then we can make the default `baseUrl` to `""` so that it is more convenient.

tlylt commented:
> > However, if the site is published via GitHub Pages, then users will need to explicitly specify baseUrl='' in this workflow to override the relative url, as PR preview is via Surge.sh, without the need for a relative URL(Doesn't seem to work with a relative URL either).
> 
> The fork-build workflow can be reused for other PR previewing platforms too right other than surge? If that is the case then maybe we can try to explore the other popular platforms to see if the default relative url is `""` or `"./repo_name"`. If most of them are `""`, then we can make the default `baseUrl` to `""` so that it is more convenient.

The fork-build workflow basically installs a specified version of MarkBind and builds the site via markbind build. I think for most static site services besides GitHub Pages, they allow a subdomain e.g. (Netlify, Vercel, & Surge). The problem with GitHub Pages is that we can't create multiple sites for one repo, hence may not work for preview purposes.

jonahtanjz commented:
In that case I think it's fine to default the `baseUrl` for this workflow to `""`. Just have to be clear in the documentation :)

tlylt commented:
I have updated the PR slightly and would like to get some opinions on:
1. In the previous discussion above I asked about the `baseUrl` option, I realized that in this case, I should just remove the option and force baseUrl to be `''` as there is no other allowable value when using surge.sh. Hope that is ok @jonahtanjz
2. I have added a new unpublish-preview.yml to allow for the automatic teardown of the preview site if required. It is slightly less "secure" because of the use of "pull_request_event", which allows limited access to the secrets, for workflows triggered by a fork PR. I am not sure if there's another way but it should be fine according to the detailed explanation about the risk [here](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/). It is secure because this workflow does not checkout the content of the fork PR at all. I have also added a note in the [readme](https://github.com/tlylt/markbind-reusable-workflows#unpublish-pr-preview-site) to inform the user. Perhaps can get another pair of eyes on this just to be sure.
3. Unlike Actions, there is no need to publish reusable workflows. Hence I am thinking of bringing the workflow files into `.github/workflows` in this repo to reduce maintenance overhead. Would like to have some opinion on this.

If anyone would like to conduct testing, the [usage](https://github.com/tlylt/markbind-reusable-workflows#usage) section provides sample workflows for users. Note that you will need to get and include a surge token in your repo in order for the workflow to work.

jonahtanjz commented:
> 1. In the previous discussion above I asked about the baseUrl option, I realized that in this case, I should just remove the option and force baseUrl to be `''` as there is no other allowable value when using surge.sh. Hope that is ok @jonahtanjz

Should be ok 👍 

> 3. Unlike Actions, there is no need to publish reusable workflows. Hence I am thinking of bringing the workflow files into `.github/workflows` in this repo to reduce maintenance overhead. Would like to have some opinion on this.

Would it be better to place it in the `.github/workflows` of the markbind-action repo? I feel that since this is related to "actions", it will be more organised if we put them in the same repo. @ang-zeyu and @ryoarmanda feel free to chip in as well :)

tlylt commented:
> Would it be better to place it in the `.github/workflows` of the markbind-action repo? I feel that since this is related to "actions", it will be more organised if we put them in the same repo. @ang-zeyu and @ryoarmanda feel free to chip in as well :)

My concern is [this](https://docs.github.com/en/actions/creating-actions/about-custom-actions#choosing-a-location-for-your-action)

jonahtanjz commented:
On the other hand, I feel that if we place this in the main repo, it may be a little bit more confusing since we already have a `ci.yml` there. Having another workflow there in the same folder that is not used by the main repo may confuse new developers? I feel it will be "less coupled" if we have all the files related to actions in one repo. That's just my thought :)

ang-zeyu commented:
@tlylt @jonahtanjz Apologise as I haven't had the bandwidth to keep up with this PR. In view of the deadline, _do proceed_ with the  discussions without me : )

If you need org permissions, feel free to let me know as well.

> Would it be better to place it in the .github/workflows of the markbind-action repo? I feel that since this is related to "actions", it will be more organised if we put them in the same repo. @ang-zeyu and @ryoarmanda feel free to chip in as well :)

I'll just chime in on this from an outside perspective -- since this relates to PR previewing functionalities, I agree with @jonahtanjz that we should place this in `markbind-actions` (since there are numerous mentions of pr preview there already).

On the other hand, our docs placed in MarkBind/markbind also have rather similar example workflows, this introduces some duplication with what @raysonkoh's PR-ed the previous time. Perhaps we can streamline the 2 somehow 🤔

jonahtanjz commented:
@tlylt 

Just checking in, any updates on this?

tlylt commented:
> On the other hand, our docs placed in MarkBind/markbind also have rather similar example workflows, this introduces some duplication with what @raysonkoh's PR-ed the previous time. Perhaps we can streamline the 2 somehow 🤔

Yes, this reusable workflow is based on what we have in the docs. The benefit is that if we can provide the workflows for the users, we can abstract out some common use cases and handle them(properly) for the user. This is not to say that users cannot achieve what they want without these workflows (they can simply follow the example in the docs). Admittedly it's a bit less useful and hence I am not prioritizing this PR much.... the workflow to auto teardown of preview site is not covered in the docs though...

@jonahtanjz Yes apologies for the delay, as shared by u and @ang-zeyu, I will put up a PR over at the markbind-actions repo for the workflows and update this PR when that is done. Perhaps just one question about versioning that I would like some opinions on: the current markbind-actions, is at `v2.0.0`, should I increment patch/minor/major for this update?

jonahtanjz commented:
> Perhaps just one question about versioning that I would like some opinions on: the current markbind-actions, is at v2.0.0, should I increment patch/minor/major for this update?

I believe you have mentioned that a reusable workflow do not need to be published? If that is the case, then there is no need to do a new release?

tlylt commented:
> > Perhaps just one question about versioning that I would like some opinions on: the current markbind-actions, is at v2.0.0, should I increment patch/minor/major for this update?
> 
> I believe you have mentioned that a reusable workflow do not need to be published? If that is the case, then there is no need to do a new release?

hmm that's true, was thinking as a workflow provided for others to use, might be good to be versioned. But I guess without it is fine at the moment as well.

jonahtanjz commented on docs/devGuide/githubActions/markbindReusableWorkflows.md:
```diff
@@ -0,0 +1,28 @@
+
+<div class="lead">
+
+A list of reusable workflows that help improve CI/CD pipelines with a MarkBind site. It helps users to streamline their workflow by:
```
```suggestion
A list of reusable workflows that help to improve the CI/CD pipelines of MarkBind sites. It helps users to streamline their workflow by:
```

jonahtanjz commented on docs/devGuide/githubActions/markbindReusableWorkflows.md:
```diff
@@ -0,0 +1,28 @@
+* Making it easier to setup PR preview for a MarkBind site, including PRs from forks.
+* Making it easier to unpublish a PR preview site after merge/close.
+
+The source code hosted alongside [MarkBind/markbind-action](https://github.com/MarkBind/markbind-action/tree/master/.github/workflows).
```
```suggestion
The source code is hosted alongside [MarkBind/markbind-action](https://github.com/MarkBind/markbind-action/tree/master/.github/workflows).
```

jonahtanjz reviewed (commented):
Thanks @tlylt!

Mostly looks good. Just 2 nits :)

jonahtanjz reviewed (approved):
Thanks @tlylt!

LGTM 👍
